Documentation changes associated to PR https://github.com/eclipse/che/pull/9577

davidfestal opened the pull request:
Documentation changes associated to PR https://github.com/eclipse/che/pull/9577 (Allow creating OpenShift objects under the current user account on OCP).

davidfestal committed:
First sketch of documentation changes for issue eclipse/che#8178

Signed-off-by: David Festal <dfestal@redhat.com>

davidfestal committed:
Small fixes

Signed-off-by: David Festal <dfestal@redhat.com>

davidfestal committed:
Add the instruction to install the Openshift identity provider

Signed-off-by: David Festal <dfestal@redhat.com>

davidfestal committed:
Add required image

Signed-off-by: David Festal <dfestal@redhat.com>

bmicklea reviewed (approved):
excellent stuff, thanks!

slemeur commented on src/main/pages/setup-openshift/admin-guide.md:
```diff
@@ -50,6 +50,10 @@ Advantages and disadvantages of using token, username/pass, and service account:
 
+**OpenShift-specific feature: create workspace objects in a personal namespace**
+
+When Che is deployed on OpenShift in mullti-user mode, there is an additional option that allows creating workspace objects **on behalf of the currenlty logged-in user**, instead of using the configured user described above. Refer to the [OpenShift Admin Guide](openshift-admin-guide#create-workspace-objects-in-personal-namespaces) for more details about how it works.
```
typo "multi-user"

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+
+## Create workspace objects in personal namespaces
+
+When Che is installed on Openshift in multi-user mode, it is possible to register the Openshift server into the Keycloak server as an identity provider,
```
typo OpenShift

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+When Che is installed on Openshift in multi-user mode, it is possible to register the Openshift server into the Keycloak server as an identity provider,
+in order to allow creating workspace objects in the personal OpenShift namespace of the user that is currenlty logged in Che through Keycloak.
+
+This feature is available only when Che is confirgured to create a new OpenShift namespace for every Che workspace.
```
typo: is configured

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+
+#### OpenShift identity provider registration
+
+The Keycloak Openshift identity provider is described in [this documentation](https://www.keycloak.org/docs/3.3/server_admin/topics/identity-broker/social/openshift.html).
```
typo: OpenShift

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+
+{% include image.html file="git/kc_roles.png" %}
+
+3. Then this identity provider has to be declared as an OAuth client inside Openshift. This can be done with the corresponding command:
```
typo: OpenShift

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+- the `CHE_INFRA_OPENSHIFT_PROJECT` environment variable should be set to `NULL` to ensure a new distinct OpenShift namespace is created for every started workspace.
+- the `CHE_INFRA_OPENSHIFT_OAUTH__IDENTITY__PROVIDER` environment variable should be set to the alias of the OpenShift identity provider
+registered in Keycloak. The default value is `openshift-v3`.
+
```
How to retrieve the OpenShift identity provider, if it's different from default?

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+
+#### Providing the Openshift certificate to Keycloak
+
+If the certificate used by the Openshift console is self-signed or is not trusted, then by default the Keycloak will not be able to
```
typo: OpenShift

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+- the `CHE_INFRA_OPENSHIFT_OAUTH__IDENTITY__PROVIDER` environment variable should be set to the alias of the OpenShift identity provider
+registered in Keycloak. The default value is `openshift-v3`.
+
+#### Providing the Openshift certificate to Keycloak
```
typo: OpenShift

slemeur reviewed (changes requested):
Good things. Just minor typos and one question!

bmicklea commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+To enable this feature, the administrator should:
+- register, inside Keycloak, an Openshift identity provider that will point to the Openshift console of the cluster in which the workspace resources should be created,
+- configure Che to use this Keycloak identity provider in order to retrieve the Openshift tokens of Che users.
+
```
Sorry, I know I approved this but in re-reading it I think we should be more clear on how to do these changes. Can we either write up the steps people need to do here explicitly or at least link off to specific pages of the Keycloak docs?  We need to assume the person using this doesn't know anything about keycloak, openshift or che.

bmicklea replied:
Can we remove this section then and either (a) bring the more actionable section up or (b) provide a link down. We should avoid sections in the docs that talk generally about what you can do without being actionable.

bmicklea replied:
Understood - but what value does the user get from reading what they'll have to do without an indication of how to do it?  Wouldn't it be better to just say 'to enable this feature follow the instructions in section X' or 'enabling this feature will allow you to easily use Che workspaces from a desktop IDE'

davidfestal committed:
fix typos

Signed-off-by: David Festal <dfestal@redhat.com>

davidfestal committed:
Changes accoring to comments from @bmicklea and @slemeur:

- Better explain that the Che properety should be set accoring to the
alias given in the previous step of keycloak identity provider
registration
- Mention that detailed installation steps are given below and link to
each related detailed section in the summary text.

Signed-off-by: David Festal <dfestal@redhat.com>

slemeur commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -9,3 +9,76 @@ folder: setup-openshift
+The environment variable is named `OPENSHIFT_IDENTITY_PROVIDER_CERTIFICATE`.
+
+Since adding a multi-line certificate content in a deployment configuration environment variable is not that easy, the best way is to use a secret that contains the certificate,
+and refer to it in the environment variable.
```
Should we put an example of how to do that?

bmicklea commented on src/main/pages/setup-openshift/openshift-admin-guide.md:
```diff
@@ -12,40 +12,40 @@ folder: setup-openshift
 {% include image.html file="git/kc_roles.png" %}
 
-3. Then this identity provider has to be declared as an OAuth client inside Openshift. This can be done with the corresponding command:
+3. Then this identity provider has to be declared as an OAuth client inside OpenShift. This can be done with the corresponding command:
```
Perhaps a dumb question - but would this work equally well with any other Keycloak-supported authentication mechanism?  Most organizations use a directory-based auth like LDAP/AD

bmicklea replied:
So the requirement is just that the identity provider for this needs to be the same as the identity provider for OpenShift. They could both be LDAP though, right?

davidfestal replied:
This work only provides a way to link the main Keycloak account to a secondary Openshift account, in order to get the Openshift token. It doesn't change anything to the way the users log into their main Keycloak account / identity. That's why the registered Openshift identity provider should have the `Link Only`, `Store Tokens` and `Stored Tokens Readable` options (see these options [here](https://www.keycloak.org/docs/3.4/server_admin/index.html#_general-idp-config) ).

However I don't see why other identity providers used for logging into Keycloak couldn't be used. Regarding LDAP I also don't see any reason why the [LDAP Federated user storage provider](https://www.keycloak.org/docs/3.4/server_admin/index.html#_ldap) wouldn't work, at least when the [`import ` storage mode](https://www.keycloak.org/docs/3.4/server_admin/index.html#storage-mode) is enabled.

bmicklea replied:
okay the main thing I wanted to make sure of is that if an organization is using LDAP to access OpenShift that won't break this integration. And it sounds like that's true...

davidfestal replied:
> okay the main thing I wanted to make sure of is that if an organization is using LDAP to access OpenShift that won't break this integration. And it sounds like that's true...

@bmicklea Then it seems I had not correctly understood your question: I answered about the case when users would like to connect to the Che server (=> to the Keycloak main account) through LDAP or any other identity provider.

> would this work equally well with any other Keycloak-supported authentication mechanism? - No, unless OpenShift auth is tied to ldap. We need a token from OpenShift with full user rights

@eivantsov could you explain why you say that ?

davidfestal committed:
Small typo in doc that leads to an installation error. 

Signed-off-by: David Festal <dfestal@redhat.com>
